stan/math/prim/fun/gaus_interp.hpp

```diff
@@ -0,0 +1,167 @@
+a_i, b_i such that the line between (x_i, y_i) and (x_{i+1}, y_{i+1}) is
+f(t) = a_i*t + b_i, store the a_i's and b_i's in as and bs.
+*/
+void lin_interp_coefs(int n, std::vector<double> xs, std::vector<double> ys,
+                      std::vector<double>& as, std::vector<double>& bs) {
+  // initialize size of vectors of linear coefficients
+  as.resize(n - 1);
+  bs.resize(n - 1);
+
+  // find slope and intercept between each point
+  for (int i = 0; i < n - 1; i++) {
+    as[i] = (ys[i + 1] - ys[i]) / (xs[i + 1] - xs[i]);
+    bs[i] = -xs[i] * as[i] + ys[i];
+  }
+}
+
+/*
+linear interpolation at one point, x, given the coefficients
+*/
+double lin_interp_pt(int n, vector<double> xs, vector<double> ys,
+                     vector<double> as, vector<double> bs, double x) {
+  // find interval where x lives
+  if (x <= xs[0])
+    return ys[0];
+  if (x >= xs[n - 1])
+    return ys[n - 1];
+
+  auto lb = upper_bound(xs.begin(), xs.end(), x);
+  int ind = distance(xs.begin(), lb) - 1;
+  ind = std::max(0, ind);
+
+  return as[ind] * x + bs[ind];
+}
+
+/*
+linear interpolation at a vector of points
+*/
+vector<double> lin_interp(int n, std::vector<double> xs, std::vector<double> ys,
+                          int n_new, std::vector<double> xs_new) {
+  // compute coefficients of linear interpolation
+  vector<double> as, bs;
+  lin_interp_coefs(n, xs, ys, as, bs);
+
+  // evaluate at new points
+  vector<double> ys_new;
+  for (int i = 0; i < n_new; i++) {
+    ys_new.push_back(lin_interp_pt(n, xs, ys, as, bs, xs_new[i]));
+  }
+  return ys_new;
+}
+
+/*
+given a set of points and a width for the gaussian kernel, do a convolution
+and evaluate at one point, x
+*/
+template <typename Tx>
+inline return_type_t<Tx> interp_gaus_pt(int n, std::vector<double> xs,
+                                        std::vector<double> ys,
+                                        std::vector<double> as,
+                                        std::vector<double> bs, Tx const& x,
+                                        double sig2) {
+  // extend out first and last lines for convolution
+  double sig = std::sqrt(sig2);
+  xs[0] += -10 * sig;
+  xs[n - 1] += 10 * sig;
+
+  // no need to convolve far from center of gaussian, so
+  // get lower and upper indexes for integration bounds
+  auto lb = upper_bound(xs.begin(), xs.end(), x - 10 * sig);
+  int ind_start = distance(xs.begin(), lb) - 1;
+  ind_start = std::max(0, ind_start);
+
+  auto ub = upper_bound(xs.begin(), xs.end(), x + 10 * sig);
+  int ind_end = distance(xs.begin(), ub);
+  ind_end = std::min(n - 1, ind_end);
+
+  // sum convolutions over intervals
+  using return_t = return_type_t<Tx>;
+  return_t y = 0;
+  for (int i = ind_start; i < ind_end; i++) {
+    y += conv_gaus_line(xs[i], xs[i + 1], as[i], bs[i], x, sig2);
+  }
+
+  return y;
+}
+
+/*
+find the smallest difference between successive elements in a sorted vector
+*/
+template <typename Tx>
+double min_diff(int n, std::vector<Tx> xs) {
```

Comment:
If this is only used in this file as an internal function then it should be in the internal namespace